tests/bsim/bluetooth/host/l2cap/multilink_peripheral/src/dut.c

```diff
@@ -0,0 +1,337 @@
+
+		LOG_DBG("[%s] Done sending", addr);
+	}
+}
+
+static struct test_ctx *get_ctx_from_chan(struct bt_l2cap_chan *chan)
+{
+	struct bt_l2cap_le_chan *le_chan = CONTAINER_OF(chan, struct bt_l2cap_le_chan, chan);
+	struct test_ctx *ctx = CONTAINER_OF(le_chan, struct test_ctx, le_chan);
+
+	TEST_ASSERT(PART_OF_ARRAY(contexts, ctx), "memory corruption");
+
+	return ctx;
+}
+
+static void sent_cb(struct bt_l2cap_chan *chan)
+{
+	struct test_ctx *ctx = get_ctx_from_chan(chan);
+
+	LOG_DBG("%p", chan);
+
+	ctx->sdu_count++;
+
+	continue_sending(ctx);
+}
+
+static void l2cap_chan_connected_cb(struct bt_l2cap_chan *chan)
+{
+	struct bt_l2cap_le_chan *le_chan = CONTAINER_OF(chan, struct bt_l2cap_le_chan, chan);
+
+	LOG_DBG("%p (tx mtu %d mps %d) (tx mtu %d mps %d)", chan, le_chan->tx.mtu, le_chan->tx.mps,
+		le_chan->rx.mtu, le_chan->rx.mps);
+
+	LOG_DBG("initiating SDU transfer");
+	continue_sending(get_ctx_from_chan(chan));
+}
+
+static void l2cap_chan_disconnected_cb(struct bt_l2cap_chan *chan)
+{
+	LOG_DBG("%p", chan);
+}
+
+static int recv_cb(struct bt_l2cap_chan *chan, struct net_buf *buf)
+{
+	TEST_FAIL("DUT should not receive data");
+
+	return 0;
+}
+
+static struct bt_l2cap_chan_ops ops = {
+	.connected = l2cap_chan_connected_cb,
+	.disconnected = l2cap_chan_disconnected_cb,
+	.recv = recv_cb,
+	.sent = sent_cb,
+};
+
+static int connect_l2cap_channel(struct bt_conn *conn, struct bt_l2cap_le_chan *le_chan)
+{
+	memset(le_chan, 0, sizeof(*le_chan));
+	le_chan->chan.ops = &ops;
+
+	return bt_l2cap_chan_connect(conn, &le_chan->chan, 0x0080);
+}
+
+static bool addr_in_use(bt_addr_le_t *address)
+{
+	return !bt_addr_le_eq(address, BT_ADDR_LE_ANY);
+}
+
+static struct test_ctx *alloc_ctx(void)
+{
+	for (int i = 0; i < ARRAY_SIZE(contexts); i++) {
+		struct test_ctx *context = &contexts[i];
+		struct bt_l2cap_le_chan *le_chan = &context->le_chan;
+
+		if (le_chan->state != BT_L2CAP_DISCONNECTED) {
+			continue;
+		}
+
+		if (addr_in_use(&context->peer)) {
+			continue;
+		}
+
+		memset(context, 0, sizeof(struct test_ctx));
+
+		return context;
+	}
+
+	return NULL;
+}
+
+static struct test_ctx *get_ctx_from_address(const bt_addr_le_t *address)
+{
+	for (int i = 0; i < ARRAY_SIZE(contexts); i++) {
+		struct test_ctx *context = &contexts[i];
+
+		if (bt_addr_le_eq(address, &context->peer)) {
+			return context;
+		}
+	}
+
+	return NULL;
+}
+
+static void acl_connected(struct bt_conn *conn, uint8_t err)
+{
+	const bt_addr_le_t *central = bt_conn_get_dst(conn);
+	char addr[BT_ADDR_LE_STR_LEN];
+	struct test_ctx *ctx;
+
+	bt_addr_le_to_str(central, addr, sizeof(addr));
+
+	TEST_ASSERT(err == 0, "Failed to connect to %s (0x%02x)", addr, err);
+
+	UNSET_FLAG(ADVERTISING);
+
+	LOG_DBG("[%s] Connected (conn %p)", addr, conn);
+
+	ctx = get_ctx_from_address(central);
+	if (ctx == NULL) {
+		LOG_DBG("no initialized context for %s, allocating..", addr);
+		ctx = alloc_ctx();
+		TEST_ASSERT(ctx, "please download more RAM");
```

Comment:
Got a link to where we can download more RAM?

```suggestion
		TEST_ASSERT(ctx, "Could not allocate ctx");
```

Reply:
sure https://downloadmoreram.com/